### PR TITLE
cnf-tests: add tap device tests

### PR DESCRIPTION
**Pull request by mmirecki**

### CODE REVIEW

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 1]
```diff
@@ -946,50 +1008,57 @@ func CreateSriovNetwork(sriovDevice *sriovv1.InterfaceExt, sriovNetworkName stri
...
-
-expect -f test.sh
-
-sleep INF
-`, dpdkResourceName, testpmdCommand, dpdkResourceName, runningTime)},
+			command,
+		},
 		SecurityContext: &corev1.SecurityContext{
-			RunAsUser: pointer.Int64Ptr(0),
+			Privileged: &privileged, // TODO: remove
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:1060]: we should not use privileged here.

I think using `NET_ADMIN` and `NET_RAW` should be ok

Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go]: We need NET_RAW, NET_ADMIN and MKNOD (to set up a tap device).
Removed privileged

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 2]
```diff
@@ -958,43 +1058,15 @@ func createDPDKWorkload(nodeSelector map[string]string, dpdkResourceName, testpm
...
+			command,
+		},
 		SecurityContext: &corev1.SecurityContext{
 			RunAsUser: pointer.Int64Ptr(0),
 			Capabilities: &corev1.Capabilities{
 				// Enable NET_RAW is required by mellanox nics as they are using the netdevice driver
 				// NET_RAW was removed from the default capabilities
 				// https://access.redhat.com/security/cve/cve-2020-14386
-				Add: []corev1.Capability{"IPC_LOCK", "SYS_RESOURCE", "NET_RAW"},
+				Add: []corev1.Capability{"IPC_LOCK", "SYS_RESOURCE", "NET_RAW", "NET_ADMIN", "MKNOD"},
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:1069]: please add them only for the vhost-net test. 
we want to check that for regular dpdk NET_ADMIN is not needed and want to say as low as possible on capabilities

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 3]
```diff
@@ -1397,3 +1469,29 @@ func CleanSriov() {
 	}
 	sriov.WaitStable(sriovclient)
 }
+
+func getDeviceMac(pod *corev1.Pod, device string) (string, error) {
+	getMacOutput, err := pods.ExecCommand(client.Client, *pod, []string{"ip", "link", "show", "dev", device})
+	if err != nil {
+		return "", err
+	}
+	macLine := strings.Trim(strings.Split(getMacOutput.String(), "\n")[1], " ")
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:1478]: please check the len here before you do the `[1]` to be sure we are not going to fly

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 4]
```diff
@@ -1397,3 +1469,29 @@ func CleanSriov() {
...
+	return strings.Split(strings.Trim(macLine, " "), " ")[1], nil
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:1479]: please check the len here before you do the `[1]` to be sure we are not going to fly

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 5]
```diff
@@ -1397,3 +1469,29 @@ func CleanSriov() {
...
+}
+
+func getDeviceRXBytes(pod *corev1.Pod, device string) (int, error) {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:1482]: can you please add a function comment and example for the expected output

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 6]
```diff
@@ -1397,3 +1469,29 @@ func CleanSriov() {
...
+	statsCommand := []string{"ip", "-s", "l", "show", "dev", device}
+	stats, err := pods.ExecCommand(client.Client, *pod, statsCommand)
+	if err != nil {
+		return 0, err
+	}
+	statsLines := strings.Split(stats.String(), "\n")
+	for i, line := range statsLines {
+		if strings.Contains(strings.Trim(line, " "), "RX:") {
+			nextLine := strings.Trim(statsLines[i+1], " ")
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:1491]: please check the len here before you do the `[i+1]` to be sure we are not going to fly

**Review by SchSeba — COMMENTED** [review]
Thanks for the PR I leave some comments but the main issue in the PR is that I was not able to find the place where you configure the sriovNetworkNodePolicy to mount the vhost-net device.

you can please also add a check before you start the dpdk application that /dev/vhost-net exist inside the container?

**Comment by mmirecki** [comment]
Waiting for: https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pull/394

**Comment by SchSeba** [comment]
@mmirecki everything is merged in the sriov operator, can you please rebase and check this PR.

Also please remove the mknod capability and the creation of the tun device.

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 7]
```diff
@@ -50,10 +50,16 @@ import (
...
-	SERVER_TESTPMD_COMMAND = "testpmd -l ${CPU} -w ${PCIDEVICE_OPENSHIFT_IO_%s} --iova-mode=va -- -i --portmask=0x1 --nb-cores=2 --forward-mode=mac --port-topology=loop --no-mlockall"
-	CLIENT_TESTPMD_COMMAND = "testpmd -l ${CPU} -w ${PCIDEVICE_OPENSHIFT_IO_%s} --iova-mode=va -- -i --portmask=0x1 --nb-cores=2 --eth-peer=0,ff:ff:ff:ff:ff:ff --forward-mode=txonly --no-mlockall"
+	LOG_ENTRY                 = "Accumulated forward statistics for all ports"
+	DEMO_APP_NAMESPACE        = "dpdk"
+	SERVER_TESTPMD_COMMAND    = "testpmd -l ${CPU} -w ${PCIDEVICE_OPENSHIFT_IO_%s} --iova-mode=va -- -i --portmask=0x1 --nb-cores=2 --forward-mode=mac --port-topology=loop --no-mlockall"
+	CLIENT_TESTPMD_COMMAND    = "testpmd -l ${CPU} -w ${PCIDEVICE_OPENSHIFT_IO_%s} --iova-mode=va -- -i --portmask=0x1 --nb-cores=2 --eth-peer=0,ff:ff:ff:ff:ff:ff --forward-mode=txonly --no-mlockall"
+	CREATE_TAP_DEVICE_COMMAND = `
+		mkdir -p /dev/net
+		mknod /dev/net/tun c 10 200
+		chmod 600 /dev/net/tun
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:60]: there is no need to create the device any more. just leave the `ip tuntap` command

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 8]
```diff
@@ -173,6 +179,99 @@ var _ = Describe("dpdk", func() {
...
+				var err error
+
+				command := CREATE_TAP_DEVICE_COMMAND + `
+				dpdk-testpmd --vdev net_tap0,iface=tap23 --no-pci -- -ia --forward-mode txonly
+				sleep INF
+				`
+				txDpdkWorkloadPod, err := createDPDKWorkload(nodeSelector,
+					command,
+					false,
+					[]corev1.Capability{"NET_ADMIN", "MKNOD"},
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:217]: no need for the `mknod`

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 9]
```diff
@@ -173,6 +179,99 @@ var _ = Describe("dpdk", func() {
...
+			It("Should be able to transmit packets", func() {
+				var err error
+				var out string
+				// --stats-period is used to keep the command alive once the pod is started
+				serverCommand := fmt.Sprintf(`
+%s
+dpdk-testpmd --vdev net_tap0,iface=tap23 -w ${PCIDEVICE_OPENSHIFT_IO_%s} -- --stats-period 5
+sleep INF
+				`, CREATE_TAP_DEVICE_COMMAND, strings.ToUpper(dpdkResourceName))
+				dpdkWorkloadPod, err := createDPDKWorkload(nodeSelector, serverCommand, false, []corev1.Capability{"NET_ADMIN", "MKNOD"})
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:243]: no need for the `mknod` anymore

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 10]
```diff
@@ -173,6 +179,99 @@ var _ = Describe("dpdk", func() {
...
+				Expect(err).ToNot(HaveOccurred())
+
+				clientCommand := fmt.Sprintf(`
+dpdk-testpmd -w ${PCIDEVICE_OPENSHIFT_IO_%s} -- --forward-mode txonly --eth-peer=0,%s --stats-period 5
+sleep INF
+				`, strings.ToUpper(dpdkResourceName), mac)
+				_, err = createDPDKWorkload(nodeSelector,
+					clientCommand,
+					false,
+					[]corev1.Capability{"NET_ADMIN", "MKNOD"},
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:256]: no need for the mknod anymore

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 11]
```diff
@@ -173,6 +179,99 @@ var _ = Describe("dpdk", func() {
...
+				// --stats-period is used to keep the command alive once the pod is started
+				serverCommand := fmt.Sprintf(`
+%s
+dpdk-testpmd --vdev net_tap0,iface=tap23 -w ${PCIDEVICE_OPENSHIFT_IO_%s} -- --stats-period 5
+sleep INF
+				`, CREATE_TAP_DEVICE_COMMAND, strings.ToUpper(dpdkResourceName))
+				dpdkWorkloadPod, err := createDPDKWorkload(nodeSelector, serverCommand, false, []corev1.Capability{"NET_ADMIN", "MKNOD"})
+				Expect(err).ToNot(HaveOccurred())
+
+				mac, err := getDeviceMac(dpdkWorkloadPod, "net1")
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:246]: This is not right you will not see the device if it's a intel nic.

better is to configure in the pod the mac that you want.

example:
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    k8s.v1.cni.cncf.io/networks: '[
      {
       "name": "dpdk-network-3",
       "mac": "60:00:00:00:00:01",
       "namespace": "seba"
      },
      {
       "name": "dpdk-network-4",
       "mac": "60:00:00:00:00:02",
       "namespace": "seba"
      }
    ]'
    cpu-load-balancing.crio.io: "true"
  labels:
    app: dpdk
  name: testpmd
  namespace: seba
spec:
....
```

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 12]
```diff
@@ -946,55 +1044,66 @@ func CreateSriovNetwork(sriovDevice *sriovv1.InterfaceExt, sriovNetworkName stri
...
-expect -f test.sh
-
-sleep INF
-`, dpdkResourceName, testpmdCommand, dpdkResourceName, runningTime)},
+			command,
+		},
 		SecurityContext: &corev1.SecurityContext{
 			RunAsUser: pointer.Int64Ptr(0),
 			Capabilities: &corev1.Capabilities{
 				// Enable NET_RAW is required by mellanox nics as they are using the netdevice driver
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:1103]: can you move the comment above this line (`capabilities := []corev1.Capability{"IPC_LOCK", "SYS_RESOURCE", "NET_RAW"}`)

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 13]
```diff
@@ -1397,3 +1506,42 @@ func CleanSriov() {
 	}
 	sriov.WaitStable(sriovclient)
 }
+
+func getDeviceMac(pod *corev1.Pod, device string) (string, error) {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:1510]: no need for this function lets just static allocate the mac addresses

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 14]
```diff
@@ -173,6 +179,99 @@ var _ = Describe("dpdk", func() {
...
+
+				Eventually(func() string {
+					out, err = pods.GetLog(dpdkWorkloadPod)
+					Expect(err).ToNot(HaveOccurred())
+					return out
+				}, 2*time.Minute, 1*time.Second).Should(ContainSubstring("Port statistics"),
+					"Cannot find port statistics")
+
+				By("Checking the rx output of tap device from the client DPDK application")
+				bytes, err := getDeviceRXBytes(dpdkWorkloadPod, "tap23")
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:268]: please also validate from the testpmd side like we do in the other tests

File: go.mod [hunk 15]
```diff
@@ -10,7 +10,7 @@ require (
 	github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
 	github.com/ishidawataru/sctp v0.0.0-20191218070446-00ab2ac2db07
 	github.com/k8snetworkplumbingwg/network-attachment-definition-client v0.0.0-20200626054723-37f83d1996bc
-	github.com/k8snetworkplumbingwg/sriov-network-operator v0.0.0-00010101000000-000000000000
+	github.com/k8snetworkplumbingwg/sriov-network-operator v1.0.1-0.20210923120918-14bd335c17c1
```
Comment (SchSeba) [inline go.mod:13]: please bump again the operator

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go (hunk 14 above)
Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:266]: Done

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go (hunk 13 above)
Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go]: Done

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go (hunk 12 above)
Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go]: Done

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go (hunk 11 above)
Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go]: Done

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go (hunk 10 above)
Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go]: Done

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go (hunk 9 above)
Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go]: Done

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go (hunk 8 above)
Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go]: Done

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go (hunk 7 above)
Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go]: Done

File: go.mod (hunk 15 above)
Comment (mmirecki) [reply go.mod]: Done

**Comment by SchSeba** [comment]
Hi @mmirecki please take a look on the CI lane 

```
The test [sriov] operator Custom SriovNetworkNodePolicy vhost-net and tun devices Validation Should have the vhost-net device inside the container does not have a valid description
2021/12/03 20:33:29 Failed to fill missing descriptions 'Found tests with no description'
```

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 16]
```diff
@@ -173,6 +178,107 @@ var _ = Describe("dpdk", func() {
...
+
+				By("Parsing output from the client DPDK application")
+				Eventually(func() string {
+					out, err = pods.GetLog(txDpdkWorkloadPod)
+					Expect(err).ToNot(HaveOccurred())
+					return out
+				}, 2*time.Minute, 1*time.Second).Should(ContainSubstring(LOG_ENTRY),
+					"Cannot find accumulated statistics")
+				By("Checking the tx output from the client DPDK application")
+				checkTxOnly(out)
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:229]: can you please also check the tap side here

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 17]
```diff
@@ -558,6 +656,21 @@ func checkRxTx(out string) {
...
 }
 
+// checkRx parses the output from the DPDK test application
+// and verifies that packets have passed the NIC RX queues
+func checkRx(out string) {
+	lines := strings.Split(out, "\n")
+	Expect(len(lines)).To(BeNumerically(">=", 3))
+	for i, line := range lines {
+		if strings.Contains(line, "NIC statistics for port") {
+			if getNumberOfPackets(lines[i+1], "RX") > 0 {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:666]: should you check here if i+1 exist?

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 18]
```diff
@@ -173,6 +178,107 @@ var _ = Describe("dpdk", func() {
...
+				Expect(bytes).To(BeNumerically(">", 0))
+
+				By("Parsing output from the DPDK application")
+				Eventually(func() string {
+					out, err = pods.GetLog(dpdkWorkloadPod)
+					Expect(err).ToNot(HaveOccurred())
+					return out
+				}, 8*time.Minute, 1*time.Second).Should(ContainSubstring("NIC statistics for port"),
+					"Cannot find accumulated statistics")
+				checkRx(out)
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:277]: should the function be `checkRxOnly` like `checkTxOnly`

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go [hunk 19]
```diff
@@ -74,6 +77,8 @@ var (
 	intelVendorID = "8086"
 
 	sriovNicsTable []TableEntry
+
+	dpdkWorkloadMac = "60:00:00:00:00:01"
```
[the remark this answers is not in the document]
Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go]: done

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go (hunk 16 above)
Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go:228]: done

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go (hunk 17 above)
Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go]: done

File: cnf-tests/testsuites/e2esuite/dpdk/dpdk.go (hunk 18 above)
Comment (mmirecki) [reply cnf-tests/testsuites/e2esuite/dpdk/dpdk.go]: done

**Review by SchSeba — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**